src/main/java/hudson/remoting/Engine.java

```diff
@@ -801,38 +807,115 @@ public void closeRead() throws IOException {
-                        HttpURLConnection conn = (HttpURLConnection) ping.openConnection();
-                        int status = conn.getResponseCode();
-                        conn.disconnect();
-                        if (status == 200) {
-                            break;
-                        } else {
-                            events.status(ping + " is not ready: " + status);
-                        }
-                    } catch (IOException x) {
-                        events.status(ping + " is not ready", x);
-                    }
-                }
                 reconnect();
             }
         } catch (Exception e) {
             events.error(e);
         }
     }
 
+    /**
+     * Evaluates a condition with exponential backoff until it succeeds or the timeout is reached.
+     * @param condition the condition to attempt to succeed with exponential backoff
+     * @return true if the condition succeeded, false if the condition failed and the timeout was reached
+     * @throws InterruptedException if the thread was interrupted while waiting.
+     */
+    private boolean succeedsWithRetries(SupplierThrowingException<Boolean> condition) throws InterruptedException {
+        var exponentialRetry = new ExponentialRetry(noReconnectAfter);
+        while (exponentialRetry != null) {
+            try {
+                if (condition.get()) {
+                    return true;
+                }
+            } catch (Exception x) {
+                events.status("Failed to connect: " + x.getMessage());
+            }
+            exponentialRetry = exponentialRetry.next(events);
+        }
+        return false;
+    }
+
+    @SuppressFBWarnings(
+            value = {"URLCONNECTION_SSRF_FD"},
+            justification = "url is provided by the user, and we are trying to connect to it")
+    private Boolean pingSuccessful() throws MalformedURLException {
+        // Unlike JnlpAgentEndpointResolver, we do not use $jenkins/tcpSlaveAgentListener/, as that will be
+        // a 404 if the TCP port is disabled.
+        URL ping = new URL(hudsonUrl, "login");
+        try {
+            HttpURLConnection conn = (HttpURLConnection) ping.openConnection();
+            int status = conn.getResponseCode();
+            conn.disconnect();
+            if (status == 200) {
+                return true;
+            } else {
+                events.status(ping + " is not ready: " + status);
+            }
+        } catch (IOException x) {
+            events.status(ping + " is not ready", x);
+        }
+        return false;
+    }
+
+    private static class ExponentialRetry {
+        final int factor;
+        final Instant beginning;
+        final Duration delay;
+        final Duration timeout;
+        final Duration incrementDelay;
+        final Duration maxDelay;
+
+        ExponentialRetry(Duration timeout) {
+            this(Duration.ofSeconds(0), timeout, 2, Duration.ofSeconds(1), Duration.ofSeconds(10));
```

Comment:
compare user-configurable version in #676

Reply:
jitter could be useful, but exposing these as user-level settings seem overkill to me.

Reply:
Agreed, just noting for reference.